Pod setup restart fixes and improvements (Loop issue #2117)

itsmojo opened the pull request.

itsmojo committed:
Add FirstAppear

itsmojo committed:
Alphabetize PumpManagerUI/Views files

itsmojo committed:
Add cannulaInsertionSuccessfullyStarted var to podState & pumpManager

itsmojo committed:
For an in-progress insert cannula compute a wait time on remaining bolus

itsmojo committed:
Call new resetPerPodPumpManagerState() func when starting a pod
Reset podAttachmentConfirmed state variable for each new pod
Reset all per pod PumpManagerAlerts for each new pod
Update PumpStatusHighlight to "Finish Setup" while activating a paired pod
Add comment noting that podCommState.deactivating value is never returned

itsmojo committed:
UI fixes and improvements for various pod setup restart situations
+ Rename func & enum names from pair to pairAndPrime for clarity
+ Start with pairAndPrime state when prime has not finalized
+ Gracefully handle restarts when prime or insert cannula is in progress
+ Add restart support to PairPodViewModel & View to skip button wait
+ Add restart support to InsertCannulaViewModel & View to skip button wait
+ Fix checkCannulaInsertionFinished() to use checkingInsertion state
+ Fix PodPairer.podIsActivated to properly handle a paired pod
+ Add new PodPairer.getPodCommState var to implement podIsActivated

marionbarker commented:
Successful Test using rPi DASH Simulator.

Please see OmniKit test notes as well at: https://github.com/LoopKit/OmniKit/pull/27#issuecomment-1935317938

Code Review: compared OmniBLE PR 113 with OmniKit PR 27.

Tested with rPi DASH simulator on an SE phone running iOS 17.3.
Configuration:
* LoopWorkspace-dev, commit 7340d55
* OmniBLE, dev branch commit 8f78c344
* OmniKit, main branch commit f2fc341

Refer to https://github.com/LoopKit/Loop/issues/2117#issuecomment-1932346258 for description of the test cases:

Test Case 1, 2 and 4 are no longer possible, `Deactivate` button is not available once pairing begins.

Test Case 3: returns to the Priming screen and either shows the priming progress bar advancing or shows a check mark that priming is complete.

Test Case 5: Once Cannula begins the insertion count-down, quit and restart Loop.
* Before this PR was in place, you had to slide the inserter bar before it told you insertion was complete
* Now it indicates that insertion is completed

Confirm that the autolock disable works:
* set up autolock to be 30 seconds
* Starting from the Pair Pod screen, the phone should not lock automatically until after setup has completed
* wait 30 sec between each tap
* all screens remain open past 30 sec
* when setup is completed and Loop main screen is presented, screen locks in 30 sec.

itsmojo committed:
Updated and improved resetPerPodPumpManagerState func logic
+ Use retractAlert func to handle the state.activeAlerts removal
+ Reset the state.acknowledgedTimeOffsetAlert var in case it gets used
+ Add call to resetPerPodPumpManagerState func from forgetPod()
Remove now unneeded alertsWithPendingAcknowledgment reset in forgetPod()
Simplified restart logic for insert cannula view restarts

marionbarker commented:
Reviewed most recent commit.
Tested using the latest version of this PR.

Testing Configuration: 
* iPhone 8 running iOS 16.7.5
* LoopWorkspace dev, commit b58fb45
    * Applied following OmniBLE PR: 106, 108, 109, 110 and 113
    * Added cannulaInsertionUnitsExtra = 2.0 to yield 50 sec for cannula insertion process
* pair with a real DASH pod 

* Repeated the testing previously done with rPi DASH (see comments above) and with Eros pod (see OmniKit PR 27).

This code behaved nominally when resuming following
* Quit and restart once priming had begun
   * Repeated the process during the priming and after priming completed
   * One time after resuming, had to tap Retry in the Priming screen
* Quit and restart once the insert cannula bolus had begun
   * Repeated the process during the insertion bolus and after bolus completed
   * One time after resuming, had to tap Retry in the inserting screen

~~For a novice user, is it possible to retry one more time following a restart before reporting failure to the user interface?~~

marionbarker commented:
I struck out the last sentence in the last comment because there is no quick fix. 
Emphasize in documentation to always tap Try Again a couple of times before giving up.

I believe this PR is ready to go - this is a big improvement over prior behavior for an unlikely event.

itsmojo committed:
Keep logic around resuming progress indicators in the view model
+ Remove cannulaInsertionSuccessfullyStarted var from pumpManager
+ Remove use of insertionWait variable from PodCommsSession.insertCannula()
+ Remove .restarting enum from PairPadViewModel & InsertCannulaViewModels
+ Remove restarting testing and viewModel state setting in UICoordinator
+ Remove restarting Bool var from PairPadViewModel & InsertCannulaViewModels
+ Remove handlePossibleRestart() from PairPadViewModel & InsertCannulViewModel
+ Remove use of onFirstAppear and delete now unneeded FirstAppear.swift
+ Improved PodPairer to use the pumpManager podCOmmState var
+ Update CannulaInserter to implement cannulaInsertionSuccessfullyStarted var
+ Update PairPodViewModelState priming value to use an optional
+ Have the view model init() funcs test for and handle resume cases
